draco/analysis/flagging.py

```diff
@@ -2334,46 +2334,257 @@ def process(self, data: containers.ContainerBase, mask: containers.ContainerBase
+class GeneralCombineMasks(task.SingleTask):
+    """Combine multiple masks using a user-specified logical expression.
 
-    All of the given masks must be of the same type and that type must have a `mask` dataset.
-    Any flagged value in any of the provided masks will be flagged in the output mask.
+    The input is a list of containers with `mask` datasets. Each mask is assigned
+    a variable name (`A`, `B`, `C`, ..., `Z`) in the order they appear. The logical
+    combination is defined using a Python expression involving those variables.
 
-    Assumes that a sample marked `True` is flagged.
+    For example, if `masks = [m1, m2]`, then the expression
+    `"A & ~B"` would keep values that are masked in `m1` and not in `m2`.
+
+    Attributes
+    ----------
+    expression : str
+        A Python expression combining the mask variables. Variables must be uppercase
+        letters `A`, `B`, ..., matching the order of the input masks.
+        The expression must evaluate to a boolean array of the same shape.
     """
 
+    expression = config.Property(proptype=str)
+
+    _dataset_name = "mask"
+
     def process(self, masks: list[containers.ContainerBase]):
-        """Combine the given list of masks into a single mask.
+        """Combine the given list of masks using the logical expression.
 
         Parameters
         ----------
-        masks
-            A list of containers that all have the same type. The type
-            must have a `mask` dataset.
+        masks : list of containers.ContainerBase
+            A list of containers with a `mask` dataset, all of the same type and shape.
 
         Returns
         -------
-        combined_mask
-            A combined mask such that any flagged value in any of the
-            input masks is flagged in the output mask.
+        combined_mask : containers.ContainerBase
+            A new container of the same type with the result of the logical combination.
         """
-        # Check that all types are the same.
+        if len(masks) > 26:
+            raise ValueError("Too many masks: only A-Z are supported (max 26).")
+
         if any(type(mask) is not type(masks[0]) for mask in masks[1:]):
-            raise TypeError(
-                "At least one type mismatch between masks. All masks must have the same type."
-            )
+            raise TypeError("All input masks must be of the same container type.")
 
-        # Create combined output mask
-        combined_mask = masks[0].copy()
+        for mask in masks:
+            mask.redistribute("freq")
+
+        # Assign variables A, B, C, ..., one per mask
+        namespace = {
+            chr(ord("A") + i): mask.datasets[self._dataset_name][:]
+            for i, mask in enumerate(masks)
+        }
+
+        # Evaluate the logical expression
+        self.log.info(f"Evaluating mask combination expression: '{self.expression}'")
+        result = eval(self.expression, {}, namespace)
```

Comment:
It's worth checking what's in `self.expression` before running `eval` since someone could in theory execute some arbitrary python code. This should be an easy check, since `self.expression` can only be a sequence of letters and logical symbols. 

I think you can also just omit the `globals` argument and provide `namespace` by keyword so it's more immediately obvious what's being passed to `eval` (I didn't know off the top of my head, so I had to look up the definition of `eval` to see what the extra arguments were).

Reply:
Okay I'm now validating the provided expression by comparing to a set of allowed characters and operators.  I'm also passing locals directly as a keyword argument to eval.